### PR TITLE
feat(470): add getTemplate function in templateFactory

### PR DESCRIPTION
**Pull request by minzcmu**
- add getTemplate function in templateFactory
- will return latest version of template based on name, version, label provided

### CODE REVIEW

File: test/lib/templateFactory.test.js [hunk 1]
```diff
@@ -122,4 +128,63 @@ describe('Template Factory', () => {
...
+                    name: 'testTemplateName',
+                    version: '1.0.2',
+                    labels: ['testLabel', 'otherLabel']
+                },
+                {
+                    id: 'd398fb192747c9a0124e9e5b4e6e8e841cf8c71c',
+                    name: 'testTemplateName',
+                    version: '1.0.3',
+                    labels: ['otherLabel']
+                }
```
Comment (d2lam) [inline test/lib/templateFactory.test.js]: Can we add an entry that is version `2`, to see if it filters it out correctly?
Also, let's make this in random order, since the `rangeKey` doesn't work and might not return them in this ascending order.

**Review by d2lam — CHANGES_REQUESTED** [review]

File: test/lib/templateFactory.test.js (hunk 1 above)
Comment (minzcmu) [reply test/lib/templateFactory.test.js]: addressed

File: lib/templateFactory.js [hunk 2]
```diff
@@ -52,6 +53,36 @@ class TemplateFactory extends BaseFactory {
...
+            .then((templates) => {
+                // get templates that have version prefix as config.version
+                const filtered = templates.filter(template =>
+                    template.version.startsWith(config.version));
+
+                // Sort templates by descending order
+                filtered.sort((a, b) => compareVersions(b.version, a.version));
+
+                // If no config.label, return latest one
+                if (filtered.length > 0 && !config.label) {
```
Comment (FenrirUnbound) [inline lib/templateFactory.js:76]: What if `config.label` is empty?

File: test/lib/templateFactory.test.js [hunk 3]
```diff
@@ -122,4 +123,69 @@ describe('Template Factory', () => {
...
+
+            return factory.getTemplate(config).then((model) => {
+                assert.instanceOf(model, Template);
+                Object.keys(expected).forEach((key) => {
+                    assert.strictEqual(model[key], expected[key]);
+                });
+            });
+        });
+
+        it('should get the correct template for a given config without label', () => {
```
Comment (FenrirUnbound) [inline test/lib/templateFactory.test.js:179]: There are no error cases described

**Review by FenrirUnbound — CHANGES_REQUESTED** [review]

File: lib/templateFactory.js (hunk 2 above)
Comment (FenrirUnbound) [reply lib/templateFactory.js:76]: For some odd reason, I misinterpreted config.label to be an array. As a String, my concern isn't valid.

This is cool 👍

File: test/lib/templateFactory.test.js [hunk 4]
```diff
@@ -122,4 +123,69 @@ describe('Template Factory', () => {
...
+                {
+                    id: '151c9b11e4a9a27e9e374daca6e59df37d8cf00f',
+                    name: 'testTemplateName',
+                    version: '2.0.1',
+                    labels: ['testLabel', 'otherLabel']
+                }
+            ];
+        });
+
+        it('should get the correct template for a given config with label', () => {
```
Comment (FenrirUnbound) [inline test/lib/templateFactory.test.js:166]: What happens in the case where I give you a label that doesn't match?

Comment (FenrirUnbound) [reply test/lib/templateFactory.test.js:166]: What happens in the edge case where `super.list` returns nothing?

Comment (minzcmu) [reply test/lib/templateFactory.test.js:166]: It will return undefined. According to the doc for find: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

File: test/lib/templateFactory.test.js (hunk 3 above)
Comment (minzcmu) [reply test/lib/templateFactory.test.js:179]: Added a test case when list resolves an empty array.

**Review by d2lam — APPROVED** [review]

**Review by FenrirUnbound — APPROVED** [review]
👍 

source 5
